File: ovstorage-c-source/examples/tutorial/05_cache.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

#include "common.hpp"

int main(int argc, char** argv)
{
    if (argc < 2 || argc > 3) {
        std::cerr << "usage: " << argv[0] << " PLUGIN_DIR [HTTP_URL]\n";
        return 2;
    }
    const std::string url =
        argc == 3 ? argv[2]
                  : "https://raw.githubusercontent.com/"
                    "NVIDIA-Omniverse/ovstorage/main/README.md";
    tutorial::Context context;
    if (!context.load(argv[1], "core") ||
        !context.load(argv[1], "cache") ||
        !context.load(argv[1], "http")) return 1;

    tutorial::TempDirectory temporary("ovstorage-cpp-05");
    if (!temporary) return 1;
    const auto& work = temporary.path();
    ovstorage::Stack stack;
    if (!tutorial::add_file(stack, context, "files", work / "files") ||
        !tutorial::add_http(stack, context, "web", url)) return 1;
    // First reproduce the routed file/HTTP graph from the preceding tutorial.
    if (!tutorial::ok("add router",
            stack.add_layer(context.registry, "routes", "router")) ||
        !tutorial::ok("route children",
            stack.set_children("routes", {"files", "web"})) ||
        !tutorial::ok("add redirect follower",
            stack.add_layer(context.registry, "redirects", "redirect_follower")) ||
        !tutorial::ok("redirect inner",
            stack.set_inner("redirects", "routes")) ||
        // Wrappers compose from the root inward. MetadataCache stores stat-like
        // results but delegates misses through redirects to the router.
        !tutorial::ok("add metadata cache",
            stack.add_layer(context.registry, "metadata", "metadata_cache")) ||
        !tutorial::ok("metadata inner",
            stack.set_inner("metadata", "redirects")) ||
        // ByteCache becomes the root so it sees reads before MetadataCache.
        //
        //     content -> metadata -> redirects -> routes -> files/web
        !tutorial::ok("add content cache",
            stack.add_layer(context.registry, "content", "byte_cache")) ||
        !tutorial::ok("content inner",
            stack.set_inner("content", "metadata"))) return 1;

    // Cache policy is Layer configuration, not an option on each operation.
    // Cached content and SQLite state use separate application-owned paths.
    if (!tutorial::ok("metadata ttl", stack.add_layer_config(
            "metadata", "ttl_seconds", ovstorage::ConfigValue::int_(60))) ||
        !tutorial::ok("content cache_root", stack.add_layer_config(
            "content", "cache_root",
            ovstorage::ConfigValue::string_((work / "content").string()))) ||
        !tutorial::ok("content state_root", stack.add_layer_config(
            "content", "state_root",
            ovstorage::ConfigValue::string_((work / "state").string()))) ||
        !tutorial::ok("Stack::set_root", stack.set_root("content"))) return 1;

    auto built = ovstorage::sync_wait(stack.build());
    if (!tutorial::ok("Stack::build", built)) return 1;
    auto storage = std::move(built).value();
    // Repeating the same call demonstrates that cache lookup and population are
    // transparent to callers: the storage API is identical with or without the
    // cache Layers.
    for (int attempt = 1; attempt <= 2; ++attempt) {
        auto read = ovstorage::sync_wait(storage.read_bytes(url));
        if (!tutorial::ok("read_bytes", read)) return 1;
        std::cout << "read " << attempt << ": "
                  << read.value().first.span().size() << " bytes\n";
    }
    return 0;
}

File: ovstorage-core/examples/python/05_cache.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
"""Add plugin-provided metadata and content caches to the routed Stack."""

from __future__ import annotations

import argparse
import asyncio
import tempfile
from pathlib import Path

import ovstorage
from ovstorage.byte_cache import ByteCache
from ovstorage.file import FileBackend
from ovstorage.metadata_cache import MetadataCache
from ovstorage.plugin import PluginBackend
from ovstorage.redirect_follower import RedirectFollower
from ovstorage.router import Router

from _common import (
    file_connection_request,
    http_connection_request,
    origin_prefix,
    plugin_registry,
)


async def run(url: str, plugin_dir: str | None) -> None:
    with tempfile.TemporaryDirectory(prefix="ovstorage-05-") as directory:
        work = Path(directory)

        # Wrappers compose from the Stack root inward. ByteCache sees reads
        # first and stores content; MetadataCache independently stores stat-like
        # results; misses continue through redirect handling and routing.
        #
        #     content-cache (root/byte cache)
        #       └── metadata-cache
        #           └── redirects
        #               └── routes
        #                   ├── files
        #                   └── web
        storage = await (
            ovstorage.Stack(root="content-cache")
            .with_registry(plugin_registry(plugin_dir, "core", "cache", "http"))
            .wrapper(
                ByteCache(
                    "content-cache",
                    "metadata-cache",
                    {
                        "cache_root": ovstorage.ConfigValue.string(str(work / "content")),
                        "state_root": ovstorage.ConfigValue.string(str(work / "state")),
                    },
                )
            )
            .wrapper(
                MetadataCache(
                    "metadata-cache",
                    "redirects",
                    {"ttl_seconds": ovstorage.ConfigValue.int_(60)},
                )
            )
            .wrapper(RedirectFollower("redirects", "routes"))
            .router(Router("routes", ["files", "web"]))
            .backend(FileBackend("files"))
            .backend(PluginBackend("http", "web"))
            .connection("files", file_connection_request(work / "files", "local files"))
            .connection("web", http_connection_request(origin_prefix(url), "public HTTP"))
            .build()
        )

        # Repeating the same bounded read demonstrates that callers do not need
        # a cache-specific API. Cache lookup and fill happen inside the Layer.
        for attempt in (1, 2):
            data, _info = await storage.read_bytes(url, max_bytes=1024 * 1024)
            print(f"read {attempt}: {len(data)} bytes")

        # Metadata has its own policy and TTL, so repeated stat operations are
        # shown separately from content reads.
        for attempt in (1, 2):
            info = await storage.stat(url)
            print(f"stat {attempt}: size={info.size}, etag={info.etag}")

        # The content and state roots are application-owned locations. Keeping
        # them separate makes cached bytes distinct from SQLite bookkeeping.
        print("content cache:", work / "content")
        print("cache state:  ", work / "state")
        # Release the cache Layer and its SQLite state before TemporaryDirectory
        # removes the files. This ordering is required on Windows.
        del storage


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "url",
        nargs="?",
        default="https://raw.githubusercontent.com/NVIDIA-Omniverse/ovstorage/main/README.md",
    )
    parser.add_argument("--plugin-dir")
    args = parser.parse_args()
    asyncio.run(run(args.url, args.plugin_dir))


if __name__ == "__main__":
    main()
